tests/msc2716_test.go

```diff
@@ -913,6 +984,92 @@ func fetchUntilMessagesResponseHas(t *testing.T, c *client.CSAPI, roomID string,
 	}
 }
 
+func paginateUntilMessageCheckOff(t *testing.T, c *client.CSAPI, roomID string, expectedEventIDs []string) {
+	t.Helper()
+	start := time.Now()
+
+	workingExpectedEventIDMap := make(map[string]string)
+	for _, expectedEventID := range expectedEventIDs {
+		workingExpectedEventIDMap[expectedEventID] = expectedEventID
+	}
+
+	var actualEventIDList []string
+	checkCounter := 0
+	messageResEnd := ""
+	generateErrorMesssageInfo := func() string {
+		i := 0
+		leftoverEventIDs := make([]string, len(workingExpectedEventIDMap))
+		for eventID := range workingExpectedEventIDMap {
+			leftoverEventIDs[i] = eventID
+			i++
+		}
+
+		return fmt.Sprintf("Called /messages %d times but only found %d/%d expected messages. Leftover messages we expected (%d): %s. We saw %d events over all of the API calls: %s",
+			checkCounter,
+			len(expectedEventIDs)-len(leftoverEventIDs),
+			len(expectedEventIDs),
+			len(leftoverEventIDs),
+			leftoverEventIDs,
+			len(actualEventIDList),
+			actualEventIDList,
+		)
+	}
+
+	for {
+		if time.Since(start) > 200*c.SyncUntilTimeout {
```

Comment:
This timeout is currently very large to accommodate the long ~20s `/messages` durations. We also have to make that request 11 times during the test which adds up very fast.

Synapse really has to chug for those requests 👹 and ideally wouldn't have to modify this at all.

I would need to look into optimizing Synapse to make this fast which we should probably do anyway as this is painfully slow.

Reply:
This is currently set to a whopping 1000s by default https://github.com/matrix-org/complement/blob/136fd601570c43fbfb1cb10647884dff2338f4b4/internal/docker/deployment.go#L57 which seems excessive even given this.

Reply:
@kegsay Sorry this wasn't clear as undrafting indicates "marked this pull request as ready for review" but I didn't assign you this one yet specifically because of this problem.

The test is good to go and `numBatches` could be switched back to `11` and shipped but want to make it actually acceptable time-wise to run against Synapse before merging.

I used `numBatches=2` during testing because it's much faster to see results while developing.

Reply:
> @MadLittleMods what is the status of this PR? It's still marked as a Draft.
>
> https://github.com/matrix-org/complement/pull/214#issuecomment-1123426080

Status is still the same as the last update in this thread. It's too slow on Synapse for me to be comfortable merging it yet so runs for everyone.

Reply:
In terms of optimizing Synapse to make this test viable to run time-wise, I'm a bit blocked on a race condition in some recent code, https://github.com/matrix-org/synapse/pull/12394#discussion_r872012615

Reply:
> In terms of optimizing Synapse to make this test viable to run time-wise, I'm a bit blocked on a race condition in some recent code, [matrix-org/synapse#12394 (comment)](https://github.com/matrix-org/synapse/pull/12394#discussion_r872012615) -> [matrix-org/synapse#12646](https://github.com/matrix-org/synapse/issues/12646)

Now https://github.com/matrix-org/synapse/issues/12988 I think

Reply:
Made progress on optimizing Synapse:

 - Saves 10s per `/messages` call: https://github.com/matrix-org/synapse/pull/13863 
 - Still need to figure out https://github.com/matrix-org/synapse/pull/13864